src/TestCase.ts

```diff
@@ -0,0 +1,241 @@
+  return {
+    start: serializePosition(range.start),
+    end: serializePosition(range.end),
+  };
+}
+
+export function serializeSelection(selection: Selection): SerializedSelection {
+  return {
+    active: serializePosition(selection.active),
+    anchor: serializePosition(selection.anchor),
+  };
+}
+
+export function serializePosition(position: Position): SerializedPosition {
+  return { line: position.line, character: position.character };
+}
+
+type TestCaseCommand = {
+  actionName: ActionType;
+  partialTargets: PartialTarget[];
+  extraArgs: any[];
+};
+
+type TestCaseContext = {
+  talonCommand: string;
+  thatMark: ThatMark;
+  targets: Target[];
+  navigationMap: NavigationMap;
+};
+
+type TestCaseSnapshot = {
+  document: string;
+  clipboard: string;
+  visibleRanges: SerializedRange[];
+  selections: SerializedSelection[];
+  thatMark: SerializedSelection[] | null;
+};
+
+type DecorationRanges = { [coloredSymbol: string]: SerializedRange };
+
+export type TestCaseFixture = {
+  talonCommand: string;
+  command: TestCaseCommand;
+  targets: Target[];
+  languageId: string;
+  marks: DecorationRanges;
+  initialState: TestCaseSnapshot;
+  finalState: TestCaseSnapshot;
+  returnValue: any;
+};
+
+export default class TestCase {
+  talonCommand: string;
+  command: TestCaseCommand;
+  languageId: string;
+  targets: Target[];
+  marks: DecorationRanges;
+  context: TestCaseContext;
+  initialState: TestCaseSnapshot | null = null;
+  finalState: TestCaseSnapshot | null = null;
+  returnValue: any = null;
+
+  constructor(command: TestCaseCommand, context: TestCaseContext) {
+    const activeEditor = vscode.window.activeTextEditor!;
+    const { navigationMap, targets, talonCommand } = context;
+
+    this.talonCommand = talonCommand;
+    this.command = command;
+    this.languageId = activeEditor.document.languageId;
+    this.marks = this.extractTargetedDecorations(targets, navigationMap);
+    this.targets = targets;
+    this.context = context;
+  }
+
+  extractPrimitiveTargetKeys(...targets: PrimitiveTarget[]) {
+    const keys: string[] = [];
+    targets.forEach((target) => {
+      if (target.mark.type === "decoratedSymbol") {
+        const { character, symbolColor } = target.mark;
+        keys.push(NavigationMap.getKey(symbolColor, character));
+      }
+    });
+    return keys;
+  }
+
+  extractTargetKeys(target: Target): string[] {
+    switch (target.type) {
+      case "primitive":
+        return this.extractPrimitiveTargetKeys(target);
+
+      case "list":
+        return target.elements.map(this.extractTargetKeys, this).flat();
+
+      case "range":
+        return this.extractPrimitiveTargetKeys(target.start, target.end);
+
+      default:
+        return [];
+    }
+  }
+
+  extractTargetedDecorations(targets: Target[], navigationMap: NavigationMap) {
+    if (!navigationMap) {
+      return {};
+    }
+
+    const decorationRanges = navigationMap.serializeRanges();
```

Comment:
Seems wasteful / not quite the right separation of concerns to have navigation map dump everything.  Might be better to do `getToken` directly for what you need

Reply:
You're right, that would feel better. I did it this way because internal map is private, so that the only way to get all the marks out is to try every color/character combo with `getToken`, which also didn't feel right.